piker/brokers/kucoin.py

```diff
@@ -0,0 +1,668 @@
+    bestAskSize: float
+    bestBid: float
+    bestBidSize: float
+    price: float
+    sequence: float
+    size: float
+    time: float
+
+
+class KucoinL2(Struct, frozen=True):
+    '''
+    Real-time L2 order book format
+
+    '''
+
+    asks: list[list[float]]
+    bids: list[list[float]]
+    timestamp: float
+
+
+class KucoinMsg(Struct, frozen=True):
+    type: str
+    topic: str
+    subject: str
+    data: list[KucoinTrade | KucoinL2]
+
+
+class BrokerConfig(Struct, frozen=True):
+    key_id: str
+    key_secret: str
+    key_passphrase: str
+
+
+def get_config() -> BrokerConfig | None:
+    conf, path = config.load()
+
+    section = conf.get('kucoin')
+
+    if section is None:
+        log.warning('No config section found for kucoin in config')
+        return None
+
+    return BrokerConfig(**section)
+
+
+class Client:
+    def __init__(self) -> None:
+        self._pairs: dict[str, KucoinMktPair] = {}
+        self._bars: list[list[float]] = []
+        self._key_id: str
+        self._key_secret: str
+        self._key_passphrase: str
+        self._authenticated: bool = False
+
+        config: BrokerConfig | None = get_config()
+
+        if config and config.key_id and config.key_secret and config.key_passphrase:
+            self._authenticated = True
+            self._key_id = config.key_id
+            self._key_secret = config.key_secret
+            self._key_passphrase = config.key_passphrase
+            log.info('User credentials added')
+
+    def _gen_auth_req_headers(
+        self,
+        action: Literal['POST', 'GET'],
+        endpoint: str,
+        api_v: str = 'v2',
+    ) -> dict[str, str | bytes]:
+        '''
+        Generate authenticated request headers
+        https://docs.kucoin.com/#authentication
+
+        '''
+        now = int(time.time() * 1000)
+        path = f'/api/{api_v}{endpoint}'
+        str_to_sign = str(now) + action + path
+
+        signature = base64.b64encode(
+            hmac.new(
+                self._key_secret.encode('utf-8'),
+                str_to_sign.encode('utf-8'),
+                hashlib.sha256,
+            ).digest()
+        )
+
+        passphrase = base64.b64encode(
+            hmac.new(
+                self._key_secret.encode('utf-8'),
+                self._key_passphrase.encode('utf-8'),
+                hashlib.sha256,
+            ).digest()
+        )
+
+        return {
+            'KC-API-SIGN': signature,
+            'KC-API-TIMESTAMP': str(now),
+            'KC-API-KEY': self._key_id,
+            'KC-API-PASSPHRASE': passphrase,
+            # XXX: Even if using the v1 api - this stays the same
+            'KC-API-KEY-VERSION': '2',
+        }
+
+    async def _request(
+        self,
+        action: Literal['POST', 'GET'],
+        endpoint: str,
+        api_v: str = 'v2',
+        headers: dict = {},
+    ) -> Any:
+        '''
+        Generic request wrapper for Kucoin API
+
+        '''
+        if self._authenticated:
```

Comment:
shouldn't this be the inverse?

only regen auth headers when *not* already authed?

Reply:
Let me test this - I assumed since the headers need a timestamp that you'd have to create new ones for each request. In retrospect I could totally be wrong about this and haven't tested to evaluate my assumptions.

Reply:
Ah but wait, we have to pass the HTTP action into the header... @goodboy we have to encode to action and timestamp into the headers like so:
```py
str_to_sign = str(int(time.time() * 1000)) + action + f'/api/{api_v}{endpoint}'

        signature = base64.b64encode(
            hmac.new(
                self._key_secret.encode('utf-8'),
                str_to_sign.encode('utf-8'),
                hashlib.sha256,
            ).digest()
        )
```
Worth it to try to generate some of the headers ahead of time?